c/meterpreter/source/extensions/sniffer/sniffer.c

```diff
@@ -320,54 +391,120 @@ DWORD request_sniffer_capture_start(Remote *remote, Packet *packet)
 			break;
 		}
 
-		ifh = pktsdk_interface_by_index(ifid);
-		if (ifh == NULL)
+		ifh = ifid;
+		j = &open_captures[ifid];
+
+		// the interface is already being captured
+		if (j->active)
 		{
 			result = ERROR_INVALID_PARAMETER;
 			break;
 		}
 
-		j = &open_captures[ifid];
+		name = get_interface_name_by_index(ifh);
+		dprintf("sniffer>> start_capture() name is %s\n",name);
 
-		// the interface is already being captured
-		if (j->active)
+		if(!name)
 		{
 			result = ERROR_INVALID_PARAMETER;
 			break;
 		}
+		j->pcap = pcap_open_live(name,
+			65535,
+			PCAP_OPENFLAG_PROMISCUOUS|PCAP_OPENFLAG_NOCAPTURE_LOCAL|PCAP_OPENFLAG_MAX_RESPONSIVENESS,
+			1000,
+			errbuf
+		);
+		if(!j->pcap)
+		{
+			dprintf("sniffer>> start_capture() interface acquisition failed: %s\n", errbuf);
+			result = ERROR_ACCESS_DENIED;
+			break;
+		}
+		j->capture_linktype = pcap_datalink(j->pcap);
+		dprintf("sniffer>> start_capture() linktype is %d\n", j->capture_linktype);
+		if (-1 == j->capture_linktype)
+		{
+			j->capture_linktype = 1; // force to LINKTYPE_ETHERNET in case of error
+		}
+
+		if(packet_filter)
+		{
+			struct bpf_program bpf;
+			char *add_filter;
+			char *real_filter = NULL;
+			int rc;
 
-		j->adp = AdpCreate();
-		dprintf("sniffer>> capture_start() AdpCreate: 0x%.8x", j->adp);
+			dprintf("handling packet_filter");
 
-		AdpSetConfig(j->adp, ifh);
-		hErr = AdpOpenAdapter(j->adp);
-		dprintf("sniffer>> capture_start() AdpOpenAdapter: 0x%.8x", hErr);
+			add_filter = packet_get_tlv_value_string(packet, TLV_TYPE_SNIFFER_ADDITIONAL_FILTER);
 
-		if (hErr != HNERR_OK)
+			dprintf("add_filter = %p (%s)", add_filter, add_filter ? add_filter : "");
+
+			if(add_filter)
+			{
+				dprintf("%s and (%s)", packet_filter, add_filter);
+			}
+			else
+			{
+				real_filter = _strdup(packet_filter);
+			}
+
+			dprintf("the real filter string we'll be using is '%s'", real_filter);
+
+			rc = pcap_compile(j->pcap, &bpf, real_filter, 1, 0);

```

Comment:
Could be compiling a NULL filter here if the user had provided a TLV_TYPE_SNIFFER_ADDITIONAL_FILTER above...